### ISSUE TITLE
Use a build server for automated testing?

### ISSUE THREAD

**Issue by Hyphen-ated** [boppreh/keyboard#46]
This library advertises python 2 compatibility but it looks like the latest release had zero testing on python 2, since it doesn't work at all.

I tried setting up travis-ci to run keyboard's test suite on python 2 and 3, and got it working pretty easily. Check it out: https://github.com/Hyphenated-integrations/keyboard/commits/master

https://travis-ci.org/Hyphenated-integrations/keyboard

When I revert the recent fix to the py2 issue, the py2 travis build fails! Cool! I don't know why it times out rather than giving an error message, but that's not a huge deal. It still means the code is broken. 

travis-ci is free for open source projects, so: consider doing this officially to help make fewer bad releases?

**Comment by boppreh** [comment]
You are absolutely correct. I dropped the ball badly.

I like automated build tests, but I need something stronger. Releases with failing tests shouldn't happen at all, so the solution would be Travis plus a blocking git hook. However, as you have noticed, I rather not have extra dependencies. Fortunately, I had already automated the release process, so adding a testing step is rather easy.

For standardization I removed the (`make_readme.sh` and `make_tests.sh` files and replaced them with a single Makefile. And released a new version with the fix that was already in master.

Do you think Travis is still valuable in this case?

**Comment by Hyphen-ated** [comment]
Yeah I definitely think it's still valuable. You don't want to manually run your test suite on each supported version of python, right? I assume you're mainly using python 3 yourself, so it doesn't look like the new stuff you committed would have caught the python 2 error that went out. 

You can even have travis build for a large list of specific python releases. Like you certainly don't want to manually run your tests on FIFTEEN different pythons, but travis could do that easily if you want to make sure you don't break on any of those versions.

A hook to block the release if the travis build fails is nice to have, but it's not really necessary. You can just take a look at it yourself before a release as a sanity check. In fact, it probably wouldn't even get that far, because the little red X shown in the github commit history UI would probably make someone notice it shortly after whatever commit introduces a problem. you can also set up email notifications for when a build starts failing, in case you never ever look at the github history UI.

**Comment by boppreh** [comment]
> You don't want to manually run your test suite on each supported version of python, right? 

That's exactly what I'm doing, modulo backwards compatibility of python versions. Currently it tests on Python2.7 and Python3.6, and I also have 3.3 and 3.5 installed. I think 2.7, 3.3 and 3.6 (or latest) should be enough. 

> Like you certainly don't want to manually run your tests on FIFTEEN different pythons, but travis could do that easily if you want to make sure you don't break on any of those versions.

I don't see a realistic way for e.g. work on 3.6 and 3.3 but fail on 3.5. 

> A hook to block the release if the travis build fails is nice to have, but it's not really necessary.

The entire release process is automated and done on the command line. From changelog to Readme generation, PyPI upload and Github release (and now tests). I only visit Github to reply to issues, and an asynchronous email would come too late. By the way, there a LOT of steps to release a version. I used to do this manually on another project, but it was a huge source of problems. I don't want to go back :(

The problem at hand is how to avoid mishappen releases. A non blocking Travis build goes only part of the way.

However it would be a good "nice to have" addition, providing a history of build versions and working as sanity check. 

I'll think about it.

**Comment by Hyphen-ated** [comment]
> I only visit Github to reply to issues, and an asynchronous email would come too late.

I don't understand the "too late" part -- do you keep all your commits local and only push them to github as part of a release? Travis watches the github repo and triggers itself every time there are new commits. Emails about build breakages would arrive minutes after those commits are on github

**Comment by glitchassassin** [comment]
I think with Travis CI you can run the tests on each commit, but only build & deploy when you push a new tag. I know AppVeyor supports this in Windows.

Then you'd just have to make sure the tests pass, set a new version, and let it roll out.

**Comment by boppreh** [comment]
> do you keep all your commits local and only push them to github as part of a release? 

Not always, but often, yes. It's one of the reasons why git was created on the first place. 

Anyway, I'll probably add Travis. The peace of mind and sanity check is enough reason for me.

Testing the backends is a related,  but harder problem. Nobody wants their mouse to wiggle by itself during an installation.
